rust-runtime/aws-smithy-types/src/timeout/config.rs

```diff
@@ -89,30 +120,154 @@ impl Config {
-    /// // A never set a connect timeout so B's value was used
-    /// assert_eq!(timeout_config.api.call_attempt_timeout(), TriState::Set(Duration::from_secs(3)));
+    /// assert_eq!(timeout_config.connect_timeout(), Some(Duration::from_secs(3)));
+    /// // A never set an operation timeout so B's value is used
+    /// assert_eq!(timeout_config.operation_timeout(), Some(Duration::from_secs(3)));
     /// ```
     pub fn take_unset_from(self, other: Self) -> Self {
         Self {
-            api: self.api.take_unset_from(other.api),
-            http: self.http.take_unset_from(other.http),
-            tcp: self.tcp.take_unset_from(other.tcp),
+            connect_timeout: self.connect_timeout.or(other.connect_timeout),
+            read_timeout: self.read_timeout.or(other.read_timeout),
+            operation_timeout: self.operation_timeout.or(other.operation_timeout),
+            operation_attempt_timeout: self
+                .operation_attempt_timeout
+                .or(other.operation_attempt_timeout),
         }
     }
 
-    /// Returns true if any of the possible timeouts are set
+    /// Builds a `TimeoutConfig`.
+    pub fn build(self) -> TimeoutConfig {
+        TimeoutConfig {
+            connect_timeout: self.connect_timeout,
+            read_timeout: self.read_timeout,
+            operation_timeout: self.operation_timeout,
+            operation_attempt_timeout: self.operation_attempt_timeout,
+        }
+    }
+}
+
+impl From<TimeoutConfig> for TimeoutConfigBuilder {
+    fn from(timeout_config: TimeoutConfig) -> Self {
+        TimeoutConfigBuilder {
+            connect_timeout: timeout_config.connect_timeout,
+            read_timeout: timeout_config.read_timeout,
+            operation_timeout: timeout_config.operation_timeout,
+            operation_attempt_timeout: timeout_config.operation_attempt_timeout,
+        }
+    }
+}
+
+/// Top-level configuration for timeouts
+///
+/// # Example
+///
+/// ```rust
+/// # use std::time::Duration;
+///
+/// # fn main() {
+/// use aws_smithy_types::timeout::TimeoutConfig;
+///
+/// let timeout_config = TimeoutConfig::builder()
+///     .operation_timeout(Duration::from_secs(30))
+///     .operation_attempt_timeout(Duration::from_secs(10))
+///     .connect_timeout(Duration::from_secs(3))
+///     .build();
+///
+/// assert_eq!(
+///     timeout_config.operation_timeout(),
+///     Some(Duration::from_secs(30))
+/// );
+/// assert_eq!(
+///     timeout_config.operation_attempt_timeout(),
+///     Some(Duration::from_secs(10))
+/// );
+/// assert_eq!(
+///     timeout_config.connect_timeout(),
+///     Some(Duration::from_secs(3))
+/// );
+/// # }
+/// ```
+#[non_exhaustive]
+#[derive(Clone, PartialEq, Debug)]
+pub struct TimeoutConfig {
+    connect_timeout: Option<Duration>,
+    read_timeout: Option<Duration>,
+    operation_timeout: Option<Duration>,
+    operation_attempt_timeout: Option<Duration>,
+}
```

Comment:
I don't have any suggestions here, but is it weird when the builder and the built struct are structurally identical? It feels weird. It feels good to uphold the pattern of having a separate builder for forwards compat reasons, so idk.

Reply:
I added the builder both to make `TimeoutConfig` consistent with `RetryConfig` as well as for forwards compatibility.